hudi-client/hudi-client-common/src/main/java/org/apache/hudi/common/table/log/HoodieFileSliceReader.java

```diff
@@ -20,62 +20,33 @@
 package org.apache.hudi.common.table.log;
 
 import org.apache.hudi.common.model.HoodieRecord;
-import org.apache.hudi.common.model.HoodieRecordPayload;
 import org.apache.hudi.common.util.Option;
-import org.apache.hudi.common.util.SpillableMapUtils;
 import org.apache.hudi.common.util.collection.Pair;
-import org.apache.hudi.config.HoodiePayloadConfig;
-import org.apache.hudi.exception.HoodieIOException;
 import org.apache.hudi.io.storage.HoodieFileReader;
 
 import org.apache.avro.Schema;
-import org.apache.avro.generic.GenericRecord;
 
 import java.io.IOException;
 import java.util.Iterator;
-import java.util.stream.StreamSupport;
+import java.util.Properties;
 
 /**
  * Reads records from base file and merges any updates from log files and provides iterable over all records in the file slice.
  */
-public class HoodieFileSliceReader<T extends HoodieRecordPayload> implements Iterator<HoodieRecord<T>> {
+public class HoodieFileSliceReader<T> implements Iterator<HoodieRecord<T>> {
+
   private final Iterator<HoodieRecord<T>> recordsIterator;
 
   public static HoodieFileSliceReader getFileSliceReader(
-      Option<HoodieFileReader> baseFileReader, HoodieMergedLogRecordScanner scanner, Schema schema, String payloadClass,
-      String preCombineField, Option<Pair<String, String>> simpleKeyGenFieldsOpt) throws IOException {
+      Option<HoodieFileReader> baseFileReader, HoodieMergedLogRecordScanner scanner, Schema schema, Properties props, Option<Pair<String, String>> simpleKeyGenFieldsOpt) throws IOException {
     if (baseFileReader.isPresent()) {
-      Iterator baseIterator = baseFileReader.get().getRecordIterator(schema);
+      Iterator<HoodieRecord> baseIterator = baseFileReader.get().getRecordIterator(schema);
       while (baseIterator.hasNext()) {
-        GenericRecord record = (GenericRecord) baseIterator.next();
-        HoodieRecord<? extends HoodieRecordPayload> hoodieRecord = transform(
-            record, scanner, payloadClass, preCombineField, simpleKeyGenFieldsOpt);
-        scanner.processNextRecord(hoodieRecord);
+        scanner.processNextRecord(baseIterator.next().getKeyWithParams(schema, props,
```

Comment:
Suggested transition doesn't make sense: 

1. Previously, file-reader was producing `GenericRecord`s hence we had `transform` method that was wrapping them into `HoodieRecord`
2. Now, we have file-reader instance that produces `HoodieRecord`, and so calling the method `getKeyWithParams` that returns another `HoodieRecord` doesn't really make sense.

Can you please explain why do we need it and why can't we return proper record from the file-reader in the first place?

Reply:
For old records read from FileReader and passing to HoodieMergeHanle, we do not need to extract HoodieKey. 
HoodieMergeHanle use keyGeneratorOpt to extract recordKey directly.

Reply:
Yes. What you said is possible. 

One thing needs to be added, in addition to HoodieKey, getKeyWithParams func also converts avro data to HoodieRecordPayload. I think the HoodieRecord returned by FileReader should be the original data in file rather than the processed data. Because the callers of FileReader need to process the HoodieRecord differently or not need to process it. We do not need to converge the process logic to FileReader.

Reply:
It's hard to determine a good method name. Backward compatibility(Converting avro data to HoodieRecordPayload) makes the meaning of this method seem ambiguous. Do you have any good suggestions for the name of this method?

Sorry, I did not express clearing. "processed data" means "transformed data" contrary to "original data". Some callers of FileReader prefer to use  "original data" that directly read from file. Some callers of FileReader prefer to use "transformed data" processed by getKeyWithParams.

Reply:
@wzx140 let's sync up on Slack to speed up this conversation